inference_perf/apis/user_session.py

```diff
@@ -0,0 +1,76 @@
+import logging
+import asyncio
+from typing import Any, Optional
+from pydantic import ConfigDict, Field
+
+from aiohttp import ClientResponse
+from inference_perf.apis import CompletionAPIData, InferenceInfo
+from inference_perf.utils.custom_tokenizer import CustomTokenizer
+from inference_perf.config import APIConfig
+
+logger = logging.getLogger(__name__)
+
+
+class LocalUserSession:
+    user_session_id: str
+    context: str
+
+    def __init__(self, user_session_id: str, context: str = ""):
+        self.user_session_id = user_session_id
+        self.contexts = context if context else ""
+        self._current_round = 0
+        self._in_flight: asyncio.Lock = asyncio.Lock()
+        self._waiting_rounds: asyncio.Queue[asyncio.Future[bool]] = asyncio.Queue()
+
+    async def get_context(self, round: int) -> str:
+        if not self._waiting_rounds.empty() or self._in_flight.locked():
+            # entering waiting queue
+            future: asyncio.Future[bool] = asyncio.Future()
+            self._waiting_rounds.put_nowait(future)
+            await future
+        await self._in_flight.acquire()
+        self._current_round += 1
+        return self.contexts
+
+    def update_context(self, response: str) -> None:
+        self.contexts = response
+
+        if not self._waiting_rounds.empty():
+            future = self._waiting_rounds.get_nowait()
+            future.set_result(True)
+
+        self._in_flight.release()
+
+
+class UserSessionCompletionAPIData(CompletionAPIData):
```

Comment:
One thing, this also will need to support chat API which more natively supports multiturn. 

I think that is ok to be a follow up item, but it may cause some changes to the initial approach.